src/main/kotlin/kotlinx/event/AbstractEvent.kt

```diff
@@ -4,14 +4,18 @@ abstract class AbstractEvent<T> : Event<T> {
 
     @JvmSynthetic
     override operator fun plusAssign(handler: (T) -> Unit) {
-        add(handler)
+        synchronized(this) {
+            add(handler)
+        }
     }
 
     @JvmSynthetic
     override operator fun minusAssign(handler: (T) -> Unit) {
-        remove(handler)
+        synchronized(this) {
+            remove(handler)
+        }
     }
 
-    override fun invoke(data: T) = forEach { it(data) }
+    override fun invoke(data: T) = synchronized(this) { forEach { it(data) } }
```

Comment:
It's not necessary to acquire a lock on the entire event in order to iterate over it; creating a copy using `toList` should be sufficient.

```kotlin
synchronized(this) { toList() }.forEach { it(data) }
```

That said, I'm not sure if this is ideal; perhaps a sort of `CollectionEvent` should be exposed which would allow users to delegate to e.g. a [COW `ArrayList`](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/concurrent/CopyOnWriteArrayList.html) or some other thread-safe collection.

Reply:
I am neutral:
`toList()` will create a new list, for huge events this is not a good idea. But this is ideal for performence whereas acquiring a lock on the entire event will not create a new list, but performence is lesser

Reply:
I don't believe a list of event listeners would typically get large enough for this to be an issue.

I feel like it may be appropriate to leave proper locking up to the event's underlying storage type. This would remove the overhead of locking _when it is not needed_ (i.e. single-threaded UI) and allow it when required, and avoids assigning too many responsibilities to `Event`.

Reply:
Sure it's a good idea